
---
title: "Combine GWAS Results"
author: "Deniz Akdemir"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load Required Libraries

```{r load-libraries}
library(dplyr)
```

## Define File Paths

```{r define-paths}
# Define the file paths
files <- list(
  Greenhouse_leaf2 = "/Users/denizakdemir/Library/CloudStorage/GoogleDrive-deniz.akdemir.work@gmail.com/.shortcut-targets-by-id/1n5uwNs72GdTeZxmWZKLfFkgE0FIjPxVG/Akdemir_Deniz/github/GATKPIPE2/Rscripts/OldResults/GWASforGreenHouseTraitsMarch27thThinnedSnpsWithCorLeaf2/GAPIT.Association.Filter_GWAS_results.csv",
  Greenhouse_leaf3 = "/Users/denizakdemir/Library/CloudStorage/GoogleDrive-deniz.akdemir.work@gmail.com/.shortcut-targets-by-id/1n5uwNs72GdTeZxmWZKLfFkgE0FIjPxVG/Akdemir_Deniz/github/GATKPIPE2/Rscripts/OldResults/GWASforGreenHouseTraitsMarch27thThinnedSnpsWithCorLeaf3/GAPIT.Association.Filter_GWAS_results.csv",
  Endophyte_loose = "/Users/denizakdemir/Library/CloudStorage/GoogleDrive-deniz.akdemir.work@gmail.com/.shortcut-targets-by-id/1n5uwNs72GdTeZxmWZKLfFkgE0FIjPxVG/Akdemir_Deniz/github/GATKPIPE2/Rscripts/GWAS_resultsForPilarsTraitsApr25_WithFIlter_loose/GWAS_resultsForPilarsTraitsApr25_2PC/GAPIT.Association.Filter_GWAS_results.csv",
  Endophyte_stringent = "/Users/denizakdemir/Library/CloudStorage/GoogleDrive-deniz.akdemir.work@gmail.com/.shortcut-targets-by-id/1n5uwNs72GdTeZxmWZKLfFkgE0FIjPxVG/Akdemir_Deniz/github/GATKPIPE2/Rscripts/GWAS_resultsForPilarsTraitsApr25_WithFIlter/GWAS_resultsForPilarsTraitsApr25_2PC/GAPIT.Association.Filter_GWAS_results.csv"
)
```

## Read and Combine Data

```{r read-and-combine, echo=FALSE}
# Function to read data and add a source column
read_and_label <- function(file_path, label) {
  data <- read.csv(file_path)
  data$Source <- label
  return(data)
}

# Read all files and combine them into one data frame
combined_data <- bind_rows(
  read_and_label(files$Greenhouse_leaf2, "Greenhouse_leaf2"),
  read_and_label(files$Greenhouse_leaf3, "Greenhouse_leaf3"),
  read_and_label(files$Endophyte_loose, "Endophyte_loose"),
  read_and_label(files$Endophyte_stringent, "Endophyte_stringent")
)
```

## Display Combined Data

```{r display-combined-data, echo=FALSE}
# Display the combined data
head(combined_data)
```


```{r, echo=FALSE}
knitr::kable(combined_data[,-1])
```



```{r, echo=FALSE}
unique(combined_data$SNP)

load("/Users/denizakdemir/Library/CloudStorage/GoogleDrive-deniz.akdemir.work@gmail.com/.shortcut-targets-by-id/1n5uwNs72GdTeZxmWZKLfFkgE0FIjPxVG/Akdemir_Deniz/github/GATKPIPE2/Rscripts/DataforPillarApr25_filtered/DataforPillar.RData")
GenoMat[, -1]<-round(GenoMat[,-1], 1)
GenoMat[,c("Taxa",unique(combined_data$SNP))]

write.csv(GenoMat[,c("Taxa",unique(combined_data$SNP))], file="SignificantSnpsMakers.csv")

```


```{r, echo=FALSE}
# Aggregate data to list each SNP only once with corresponding traits and experiments
organized_data <- combined_data %>%
  group_by(SNP) %>%
  summarise(
    Traits = paste(unique(traits), collapse = ", "),
    Experiments = paste(unique(Source), collapse = ", ")
  ) %>%
  ungroup()

# Display the organized data
knitr::kable(organized_data)

```

